File: 2021/puzzle24.2d.py
# Advent of code day 24, part 2
# Find smallest MONAD model number

import sys

class Range(tuple):
    def __new__(cls, min, max = None):
        if max is None: max = min
        assert(isinstance(min, int) and isinstance(max, int) and min <= max)
        return super().__new__(cls, (min, max))

    def min(self):
        return self[0]

    def max(self):
        return self[1]

class Register:
    def __init__(self, index):
        assert(isinstance(index, int))
        self.index = index

    def getRange(self, registers):
        return registers[self.index]

    def setRange(self, registers, range):
        registers[self.index] = range

def div(a, b):
    """Return a / b, rounded towards zero.  This function differs from // in
    that the latter operator rounds towards negative."""

    if (a < 0) != (b < 0):
        # Result will be negative. Round towards zero by negating, rounding
        # (tawards negative), and negating again.
        return -(-a // b)
    else:
        # Result will be positive. Round towards negative == round towards
        # zero.
        return a // b

class Op:
    def __init__(self, target, source):
        self.target = target
        self.source = source

    def exec(self, registers):
        a = self.target.getRange(registers)
        source = self.source
        b = source.getRange(registers) if isinstance(source, Register) \
            else Range(source, source)
        self.target.setRange(registers, self.execImp(a, b))

class OpInp(Op):

    prefixValues = [None] # A list of known input values

    def execImp(self, a, b):
        inputNum = b.min()
        if inputNum < len(OpInp.prefixValues):
            return Range(OpInp.prefixValues[inputNum])
        else:
            return Range(1, 9)

class OpAdd(Op):

    def execImp(self, a, b):
        return Range(a.min() + b.min(), a.max() + b.max())

class OpMul(Op):

    def execImp(self, a, b):
        minA, maxA = a
        minB, maxB = b
        results = (minA * minB, minA * maxB, maxA * minB, maxA * maxB)
        return Range(min(results), max(results))

class OpDiv(Op):

    def execImp(self, a, b):
        minA, maxA = a
        minB, maxB = b
        assert(minB != 0 and maxB != 0)
        results = (div(minA, minB), div(minA, maxB),
                   div(maxA, minB), div(maxA, maxB))
        return Range(min(results), max(results))

class OpMod(Op):

    def execImp(self, a, b):
        minA, maxA = a
        minB, maxB = b
        assert(minB > 0 and maxB > 0)
        results = (minA % minB, minA % maxB, maxA % minB, maxA % maxB)
        return Range(min(results), max(results))

class OpEql(Op):

    def execImp(self, a, b):
        if a.max() < b.min() or a.min() > b.max():
            return Range(0)  # No overlap in ranges
        elif a.max() == b.min() and a.min() == b.max():
            return Range(1)  # All the same values
        else:
            return Range(0, 1)  # Could be equal or not

operations = {
    'inp' : OpInp,
    'add' : OpAdd,
    'mul' : OpMul,
    'div' : OpDiv,
    'mod' : OpMod,
    'eql' : OpEql
}

inputCount = 0  # Number of input instructions parsed

def parseInstr(instrStr):
    """Parse the specified instruction string and return the compiled `Op` object."""
    global inputCount
    opcode, *operands = instrStr.rstrip().split(' ')
    target = Register(ord(operands[0]) - ord('w'))
    operation = operations[opcode]
    if operation is OpInp:
        assert(len(operands) < 2)
        inputCount += 1
        return OpInp(target, inputCount)
    elif operands[1] in 'wxyz':
        return operation(target, Register(ord(operands[1]) - ord('w')))
    else:
        return operation(target, int(operands[1]))

def solve(startRegisters, instructions, startInstr = 0, startInpPrefix = [None]):
    """Try to find a solution where `instructions` leaves a zero value in
    `startRegisters[z]` by appending to the initial sequence of inputs specified in
    `startInpPrefix`.  If successful, returns the complete list of 14 inputs that
    solves the expression.  If unsuccessful, returns `None`.
    """

    if len(startInpPrefix) > 14:
        return startInpPrefix  # Solved!

    inpPrefix = startInpPrefix + [ 0 ]
    for i in range(1, 10):
        inpPrefix[-1] = i
        OpInp.prefixValues = inpPrefix
        registers = startRegisters.copy()
        nextGroupInstr = len(instructions)
        nextGroupRegs  = [ ]
        # Execute instructions up to and including an `OpInp` instruction
        for instIndex in range(startInstr, len(instructions)):
            instruction = instructions[instIndex]
            instruction.exec(registers)
            if isinstance(instruction, OpInp):
                # Take snapshot after `OpInp` instruction
                assert(instruction.source == len(startInpPrefix))
                nextGroupInstr = instIndex + 1
                nextGroupRegs  = registers.copy()
                break
        # Execute the remaining instructions
        for instIndex in range(nextGroupInstr, len(instructions)):
            instruction = instructions[instIndex]
            instruction.exec(registers)

        minE, maxE = registers[3]  # z register
        if minE <= 0 and 0 <= maxE:
            # Found partial solution.  Recurse down another level.
            # Nested levels can start after last input, since everything up to
            # that point is fixed.
            # if len(inpPrefix) < 9:
            #     print("\rPartial = %-15s" % ''.join(map(lambda x : str(x), inpPrefix[1:])), end='')
            solution = solve(nextGroupRegs, instructions, nextGroupInstr, inpPrefix)
            if solution: return solution  # Solved!

    return None

infile = open("puzzle24_input.txt", "r")

instructions = [ ]

for instrStr in infile:
    instructions.append(parseInstr(instrStr))

print(f'{len(instructions)} instructions processed, solving...')

solution = solve([ Range(0, 0) ] * 4, instructions)
print("\nSolution =", ''.join(map(lambda x : str(x), solution[1:])))

File: 2021/puzzle24.2d.cpp
// puzzle24.2d.cpp                                                    -*-C++-*-

#include <algorithm>
#include <array>
#include <fstream>
#include <iostream>
#include <map>
#include <string_view>
#include <variant>
#include <vector>

#include <cassert>
#include <cstdlib>

struct Range
{
    // Range of integer values.  A function returning this aggregate can be
    // used with structured bindings.

    int min;
    int max;

    constexpr Range() : min(0), max(0) { }
    explicit constexpr Range(int val) : min(val), max(val) {}
    constexpr Range(int mn, int mx) : min(mn), max(mx) { assert(mn <= mx); }
    constexpr Range(std::pair<int, int> minmax)
        : min(minmax.first), max(minmax.second) { }
};

std::ostream& operator<<(std::ostream& os, const Range& r)
{
    return os << '(' << r.min << ", " << r.max << ')';
}

constexpr int numRegs = 4;

using RegisterSet = std::array<Range, numRegs>;

class Register
{
public:
    constexpr explicit Register(int index) : m_index(index)
        { assert(0 <= index && index < numRegs); }

    constexpr Range getRange(const RegisterSet& regs) const
        { return Range(regs[m_index]); }

    void setRange(RegisterSet& regs, const Range& range) const
        { regs[m_index] = range; }

private:
    int m_index;
};

class Op
{
public:
    using ExecFunc = Range (const RegisterSet& registers, Range a, Range b);

    constexpr Op(ExecFunc* f, Register target, int source)
        : m_execFunc(f), m_target(target), m_source(source) { }
    constexpr Op(ExecFunc* f, Register target, Register source)
        : m_execFunc(f), m_target(target), m_source(source) { }

    bool isInp() const;

    void exec(RegisterSet& registers) const
    {
        Range a = m_target.getRange(registers);
        Range b;
        if (m_source.index() == 0)
        {
            int val = std::get<0>(m_source);
            b = Range(val, val);
        }
        else
        {
            b = std::get<1>(m_source).getRange(registers);
        }

        m_target.setRange(registers, m_execFunc(registers, a, b));
    //    std::cout << m_target.getRange(registers) << std::endl;
    }

private:
    // An instruction representing an opcode and two operands.
    ExecFunc*                   m_execFunc;
    Register                    m_target;
    std::variant<int, Register> m_source;
};

std::vector<int>* prefixValues_p;

Range opInp(const RegisterSet& registers, Range a, Range b)
{
    int inputNum = b.min;
    if (inputNum < prefixValues_p->size())
        return Range((*prefixValues_p)[inputNum]);
    else
        return Range(1, 9);
}

Range opAdd(const RegisterSet& registers, Range a, Range b)
{
    return Range(a.min + b.min, a.max + b.max);
}

Range opMul(const RegisterSet& registers, Range a, Range b)
{
    auto [minA, maxA] = a;
    auto [minB, maxB] = b;
    return std::minmax(
        { minA * minB, minA * maxB, maxA * minB, maxA * maxB });
}

Range opDiv(const RegisterSet& registers, Range a, Range b)
{
    auto [minA, maxA] = a;
    auto [minB, maxB] = b;
    assert(minB != 0 and maxB != 0);
    return std::minmax(
        { minA / minB, minA / maxB, maxA / minB, maxA / maxB });
}

Range opMod(const RegisterSet& registers, Range a, Range b)
{
    auto [minA, maxA] = a;
    auto [minB, maxB] = b;
    assert(minB > 0 and maxB > 0);
    return std::minmax(
        { minA % minB, minA % maxB, maxA % minB, maxA % maxB });
}

Range opEql(const RegisterSet& registers, Range a, Range b)
{
    if (a.max < b.min or a.min > b.max)
        return Range(0);     // No overlap in ranges
    else if (a.max == b.min and a.min == b.max)
        return Range(1);     // All the same values
    else
        return Range(0, 1);  // Could be equal or not
}

inline bool Op::isInp() const { return m_execFunc == opInp; }

const std::map<std::string_view, Op::ExecFunc*> operations {
    { "inp", &opInp },
    { "add", &opAdd },
    { "mul", &opMul },
    { "div", &opDiv },
    { "mod", &opMod },
    { "eql", &opEql }
};

Op parseInstr(const char* instrStr)
{
    static int inputCount = 0;

    std::string_view opcode(instrStr, 3);
    assert(' ' == instrStr[3]);
    auto operation = operations.at(opcode);
    Register target(instrStr[4] - 'w');

    if (operation == opInp)
        return Op(operation, target, inputCount++);
    else if (std::isalpha(instrStr[6]))
        return Op(operation, target, Register(instrStr[6] - 'w'));
    else
        return Op(operation, target, std::atoi(instrStr + 6));
}

void printDigits(const char* prefix, const std::vector<int>& digits)
{
    char buff[] = "              ";
    char* p = buff;
    for (int d : digits)
        *p++ = '0' + d;
    std::cout << prefix << buff << std::flush;
}

std::vector<int> solve(const RegisterSet&      startRegisters,
                       const std::vector<Op>&  instructions,
                       int                     startInstr = 0,
                       const std::vector<int>& startInpPrefix = { })
{
    if (startInpPrefix.size() >= 14)
        return startInpPrefix;

    std::vector<int> inpPrefix;
    inpPrefix.reserve(startInpPrefix.size() + 1);
    inpPrefix = startInpPrefix;
    inpPrefix.push_back(0);

    // Iterate last prefix value in range 1 to 9
    for (int i = 1; i < 10; ++i)
    {
        inpPrefix.back() = i;
        prefixValues_p = &inpPrefix;  // Must be reset every iteration

        RegisterSet registers = startRegisters;

        int nextGroupInstr = instructions.size();
        RegisterSet nextGroupRegs;

        // Execute instructions up to and including an `OpInp` instruction.
        for (int instIndex = startInstr; instIndex < instructions.size();
             ++instIndex)
        {
            auto& instruction = instructions[instIndex];
            instruction.exec(registers);
            if (instruction.isInp())
            {
                // Take snapshot after `opInp` instruction
                nextGroupInstr = instIndex + 1;
                nextGroupRegs  = registers;
                break;
            }
        }
        // Execute the remaining instructions.
        for (int instIndex = nextGroupInstr; instIndex < instructions.size();
             ++instIndex)
        {
            auto& instruction = instructions[instIndex];
            instruction.exec(registers);
        }

        auto [minE, maxE] = registers[3];  // z register
        if (minE <= 0 and 0 <= maxE)
        {
            // Found partial solution.  Recurse down another level.
            // Nested levels can start after last input, since everything up to
            // that point is fixed.
            // if (inpPrefix.size() < 8)
            //     printDigits("\rPartial = ", inpPrefix);
            auto solution = solve(nextGroupRegs, instructions, nextGroupInstr,
                                  inpPrefix);
            if (! solution.empty())
                return solution;  // Solved!
        }
    }  // end for digit values 1 to 9

    return {};
}

int main()
{
    std::ifstream infile("puzzle24_input.txt");

    std::vector<Op> instructions;
    while (infile)
    {
        char line[16];
        infile.getline(line, sizeof(line));
        if ('\0' == line[0])
            break;
        instructions.push_back(parseInstr(line));
    }

    std::cout << instructions.size() << " instructions processed. Solving..."
              << std::endl;

    auto solution = solve(RegisterSet{}, instructions);

    printDigits("\nSolution = ", solution);
}
